pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Stage.groovy

```diff
@@ -72,26 +74,27 @@ class Stage implements Serializable {
     Stage(String name, StepsBlock steps, Agent agent, PostStage post, StageConditionals when, Tools tools,
           Environment environment, Stages parallel, boolean failFast) {
         this(name, steps, agent, post, when, tools, environment, failFast, null, null,
-                (Stages) null, (Parallel) parallel != null ? new Parallel(parallel.stages) : null, null)
+                (Stages) null, (Parallel) parallel != null ? new Parallel(parallel.stages) : null, null, null)
     }
 
     @Deprecated
     Stage(String name, StepsBlock steps, Agent agent, PostStage post, StageConditionals when, Tools tools,
           Environment environment, Stages parallel, boolean failFast, StageOptions options, StageInput input) {
         this(name, steps, agent, post, when, tools, environment, failFast, options, input,
-                (Stages) null, (Parallel) parallel != null ? new Parallel(parallel.stages) : null, null)
+                (Stages) null, (Parallel) parallel != null ? new Parallel(parallel.stages) : null, null, null)
     }
 
     @Whitelisted
     Stage(String name, StepsBlock steps, Agent agent, PostStage post, StageConditionals when, Tools tools,
           Environment environment, boolean failFast, StageOptions options, StageInput input,
-          Stages stages, Parallel parallel, Matrix matrix) {
+          Stages stages, Parallel parallel, Matrix matrix, Environment preAgentEnvironment) {
         this.name = name
         this.agent = agent
         this.post = post
         this.when = when
         this.tools = tools
         this.environment = environment
+        this.preAgentEnvironment = preAgentEnvironment
```

Comment:
@abayer @dwnusbaum  
JENKINS-43911 coming back to try to thwart my dreams - this is my response.  
Without this the matrix cell values are not available at the time that agent is resolved.  :sob:

This functionality is only in the runtime for now.  If you have a better naming suggestion, we might consider exposing this a syntax change in a future change.  For now though, this get all values able to resolve based on the matrix cell values.

Reply:
So definitely no syntax support - there are too many things that could go weird and it would be verrrrrrry confusing for users to know what to use when, etc.